### PR TITLE
Token Update

### PR DESCRIPTION
**Pull request by daniel-danev-limechain**
Signed-off-by: Daniel Danev <daniel.danev@limechain.tech>

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 1]
```diff
@@ -351,7 +351,7 @@ public ResponseCodeEnum adjustBalance(AccountID aId, TokenID tId, long adjustmen
 	public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
 		final var tId = nftId.tokenId();
 		return sanityChecked(from, to, tId, token -> {
-			if (!nftsLedger.exists(nftId)) {
+			if (!nftId.isInternal() && !nftsLedger.exists(nftId)) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:354]: In the case for normal nft transfers, the NFT id will have `from`, `to` and `serialNumber` all valid. We should not include this check `!nftId.isInternal()` since `serialNumber` will not be `-1`

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 2]
```diff
@@ -364,9 +364,11 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
 				return toFreezeAndKycValidity;
 			}
 
-			final var owner = (EntityId) nftsLedger.get(nftId, OWNER);
-			if (!owner.matches(from)) {
-				return SENDER_DOES_NOT_OWN_NFT_SERIAL_NO;
+			if (!nftId.isInternal()) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:367]: in the case for normal nft transfers, the `Nft` id will have `from`, `to` and `serialNumber` all valid. We should not include this check since `serialNumber` will never be `-1`.

We have to:
- get the owner of the nft
- if it is the default entity id -> verify that `from` is equal to the `token treasury`
- if it is not the default -> verify that `owner` matches `from`

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 3]
```diff
@@ -376,20 +378,38 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
...
+				if (isKnownTreasury(to)) {
+					// sets Nft owner to 0.0.0
+					nftsLedger.set(nftId, OWNER, EntityId.MISSING_ENTITY_ID);
+					// updates token treasury
+					var merkleToken = tokens.get().getForModify(fromTokenId(nftType));
+					merkleToken.setTreasury(fromGrpcAccountId(to));
+					if (isKnownTreasury(from)) {
+						removeKnownTreasuryForToken(from, tId);
+					}
+					addKnownTreasury(to, tId);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:392]: This is already performed in the `store.update` method

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
In `PureTransferSemanticChecks` -> `validateScopedTransferSemantics` -> `hasRepeatedAccount` we can verify that there is no `nftTransfers` that have the same `accountID` for both `sender` and `receiver`

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java (hunk 1 above)
Comment (daniel-danev-limechain) [reply hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:354]: I think we actually need this check `!nftId.isInternal()` here since that code base is also used for TokenUpdate operation where `serialNumber` is -1

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java (hunk 2 above)
Comment (daniel-danev-limechain) [reply hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:367]: This code is also used for TokenUpdate operation where `serialNumber` is -1

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 4]
```diff
@@ -536,24 +560,62 @@ public boolean scheduleExists(ScheduleID id) {
...
-							.setCreationTime(nft.getCreationTime().toGrpc())
-							.setNftID(NftID.newBuilder()
-									.setTokenID(nftId.tokenId().toGrpcTokenId())
-									.setSerialNumber(nftId.serialNumber())
-									.build())
-							.setMetadata(ByteString.copyFrom(nft.getMetadata()))
-							.build()
-			);
-		});
+		var uniqueTokenAccountCount = uniqueTokenAccountOwnerships.get().getCount(fromGrpcAccountId(aid));
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:563]: I think that it would be better if have another method `computeAssociationRanges` that would return the `start/end` indexes for both the 2 FCOTMR

That way the code would be cleaner imo. If there is no need to query a map we could return `0, 0` for start/end. 

If we have that the logic afterwards will be very clean

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 5]
```diff
@@ -536,24 +560,62 @@ public boolean scheduleExists(ScheduleID id) {
...
+				var nft = uniqueTokensMap.get(nftId);
+				nftInfos.add(
+						buildTokenNft(aid, nft, nftId)
+				);
+			});
+		} else if (uniqueTokenAccountCount <= start) {
+			var tokens = tokenStore.listOfTokensServed(aid);
+			for (var tid : tokens) {
+				uniqueTokenTreasuryOwnerships.get()
+						.get(fromGrpcTokenId(tid), (int) start, (int) end).forEachRemaining(nftId -> {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:577]: We shouldn't query the `start` here. If the `start = 50` and the `uniqueTokenAccountOwnerships` has `40`. The start should be `10` for the `uniqueTokenTreasuryOwnerships`

File: hedera-node/src/main/java/com/hedera/services/ledger/HederaLedger.java [hunk 6]
```diff
@@ -229,19 +229,19 @@ public TransferList netTransfersInTxn() {
...
-					var uniqueTransfersHere = uniqueTokenTransfers.get(token);
-					if (uniqueTransfersHere != null) {
+					final var fungibleTransfersHere = netTokenTransfers.get(token);
+					if (fungibleTransfersHere != null) {
+						purgeZeroAdjustments(fungibleTransfersHere);
 						all.add(TokenTransferList.newBuilder()
 								.setToken(token)
-								.addAllNftTransfers(uniqueTransfersHere.getNftTransfersList())
+								.addAllTransfers(fungibleTransfersHere.getAccountAmountsList())
 								.build());
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/ledger/HederaLedger.java:245]: I am not sure why we need to change this logic. Maybe a mistake from the merge?

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 7]
```diff
@@ -311,7 +314,11 @@ public void persistToken(Token token) {
 						new EntityId(uniqueToken.getOwner()), uniqueToken.getMetadata(), uniqueToken.getCreationTime());
 				currentUniqueTokens.put(merkleUniqueTokenId, merkleUniqueToken);
 				currentUniqueTokenAssociations.associate(new EntityId(uniqueToken.getTokenId()), merkleUniqueTokenId);
-				currentUniqueOwnershipAssociations.associate(treasury, merkleUniqueTokenId);
+				if (uniqueToken.getOwner().equals(token.getTreasury().getId())) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:317]: When we are minting, we are always minting to the treasury. No need to add the `if`

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 8]
```diff
@@ -322,7 +329,11 @@ public void persistToken(Token token) {
 				final var accountId = new EntityId(uniqueToken.getOwner());
 				currentUniqueTokens.remove(merkleUniqueTokenId);
 				currentUniqueTokenAssociations.disassociate(new EntityId(uniqueToken.getTokenId()), merkleUniqueTokenId);
-				currentUniqueOwnershipAssociations.disassociate(accountId, merkleUniqueTokenId);
+				if (uniqueToken.getOwner().equals(token.getTreasury().getId())) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:332]: We can add inline docs do specify that in the first case we are going in in `burn` and the second  case is from wipe (going to the else)

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 9]
```diff
@@ -333,27 +337,80 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
 			final var fromThisNftsOwned = (long) tokenRelsLedger.get(fromRel, TOKEN_BALANCE);
 			final var toNftsOwned = (long) accountsLedger.get(to, NUM_NFTS_OWNED);
 			final var toThisNftsOwned = (long) tokenRelsLedger.get(asTokenRel(to, nftType), TOKEN_BALANCE);
-			nftsLedger.set(nftId, OWNER, EntityId.fromGrpcAccountId(to));
+
+			if (isKnownTreasury(to)) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:341]: This is not correct. `to` can be a treasury but it might not be treasury for this specific token that is being transferred

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 10]
```diff
@@ -333,27 +337,80 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
...
-					merkleUniqueTokenId);
+			if (!isTreasuryForToken(to, tId)) {
+				this.uniqueOwnershipAssociations.get().associate(
+						fromGrpcAccountId(to),
+						merkleUniqueTokenId);
+			} else {
+				this.uniqueOwnershipTreasuryAssociations.get().associate(
+						fromGrpcTokenId(tId),
+						merkleUniqueTokenId);
+			}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:371]: nit: I think that it would be better if it was not `!` but rather `if(isTreasuryForToken(to, tId)`. It is easier to read imo

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 11]
```diff
@@ -333,27 +337,80 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
...
 
 			hederaLedger.updateOwnershipChanges(nftId, from, to);
 
 			return OK;
 		});
 	}
 
+	@Override
+	public ResponseCodeEnum changeOwnerWildCard(NftId nftId, AccountID from, AccountID to) {
+		{
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:381]: Why do we need the `{` ?

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 12]
```diff
@@ -333,27 +337,80 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
...
+				final var fromRel = asTokenRel(from, nftType);
+				final var toRel = asTokenRel(to, nftType);
+				final var fromNftsOwned = (long) accountsLedger.get(from, NUM_NFTS_OWNED);
+				final var fromThisNftsOwned = (long) tokenRelsLedger.get(fromRel, TOKEN_BALANCE);
+				final var toNftsOwned = (long) accountsLedger.get(to, NUM_NFTS_OWNED);
+				final var toThisNftsOwned = (long) tokenRelsLedger.get(toRel, TOKEN_BALANCE);
+
+				accountsLedger.set(from, NUM_NFTS_OWNED, fromNftsOwned - fromThisNftsOwned);
+				accountsLedger.set(to, NUM_NFTS_OWNED, toNftsOwned + fromThisNftsOwned);
+				tokenRelsLedger.set(fromRel, TOKEN_BALANCE, fromNftsOwned - fromThisNftsOwned);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:404]: The new `Token Balance` for the previous owner should be 0 since we are transferring all NFTS

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 13]
```diff
@@ -333,27 +337,80 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
...
+
+				hederaLedger.updateOwnershipChanges(nftId, from, to);
+
+				return OK;
+			});
+		}
+	}
+
 	@Override
 	public ResponseCodeEnum wipe(AccountID aId, TokenID tId, long amount, boolean skipKeyCheck) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:415]: Wipe can be deleted from this store. We already implemented it in the new model

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 14]
```diff
@@ -734,6 +791,14 @@ public ResponseCodeEnum update(TokenUpdateTransactionBody changes, long now) {
...
 			}
 			if (changes.hasTreasury() && !changes.getTreasury().equals(token.treasury().toGrpcAccountId())) {
+				if (token.tokenType().equals(TokenType.NON_FUNGIBLE_UNIQUE)) {
+					var relationship = asTokenRel(changes.getTreasury(), tId);
+					long balance = (long) tokenRelsLedger.get(relationship, TOKEN_BALANCE);
+					if (balance != 0) {
+						appliedValidity.set(NOT_SUPPORTED);
+						return;
+					}
+				}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:801]: I think that this check must be performed before line `750`. All validations are there and from line 750 down we are applying the new updates. If we revert here we won't revert the token changes

File: hedera-node/src/test/java/com/hedera/services/state/initialization/ViewBuilderTest.java [hunk 15]
```diff
@@ -63,18 +67,31 @@ public static void assertIsTheExpectedUtao(FCOneToManyRelation<EntityId, MerkleU
...
+		given(tokenStore.isTreasuryForToken(any(), any())).willReturn(false);
 
 		// when:
-		ViewBuilder.rebuildUniqueTokenViews(someUniqueTokens(), actualUta, actualUtao);
+		ViewBuilder.rebuildUniqueTokenViews(tokenStore, someUniqueTokens(), actualUta, actualUtao, actualTUtao);
 
 		// then:
 		assertIsTheExpectedUta(actualUta);
 		assertIsTheExpectedUtao(actualUtao);
+//		assertIsTheExpectedTUtao(actualTUtao); TODO case with ownership treasury
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/state/initialization/ViewBuilderTest.java:94]: What is the reason behind commenting this line

File: hedera-node/src/test/java/com/hedera/services/store/TypedTokenStoreTest.java [hunk 16]
```diff
@@ -297,7 +300,6 @@ void savesTokenAsExpected() {
 		verify(uniqueTokens).remove(expectedPastUniqTokenId);
 		verify(uniqueTokenAssociations).associate(new EntityId(modelToken.getId()), expectedNewUniqTokenId);
 		verify(uniqueTokenAssociations).disassociate(new EntityId(modelToken.getId()), expectedPastUniqTokenId);
-		verify(uniqueTokenOwnerships).associate(treasuryId, expectedNewUniqTokenId);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/store/TypedTokenStoreTest.java:300]: We have to verify that the new FCOTMR was called with associate. In general we should have code coverage on the new feature

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java [hunk 17]
```diff
@@ -496,6 +502,51 @@ public HapiApiSpec uniqueTokenTxnAccountBalance() {
...
+								.tokenType(TokenType.NON_FUNGIBLE_UNIQUE)
+								.initialSupply(0)
+								.supplyKey("supplyKeyA")
+								.treasury(TOKEN_TREASURY),
+						mintToken(A_TOKEN, List.of(ByteString.copyFromUtf8("memo"))),
+						tokenAssociate("newTreasury", A_TOKEN)
+				).when(
+						cryptoTransfer(
+								movingUnique(1, A_TOKEN).between(TOKEN_TREASURY, "newTreasury")
+						).signedBy("signingKeyTreasury", "signingKeyNewTreasury", DEFAULT_PAYER).via("cryptoTransferTxn"),
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java:523]: You dont need the `signingKeyNewTreasury`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java [hunk 18]
```diff
@@ -496,6 +502,51 @@ public HapiApiSpec uniqueTokenTxnAccountBalance() {
...
+						cryptoTransfer(
+								movingUnique(1, A_TOKEN).between("newTreasury", TOKEN_TREASURY)
+						).signedBy("signingKeyTreasury", "signingKeyNewTreasury", DEFAULT_PAYER).via("cryptoTransferTxn")
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java:526]: I am not sure why we are changing the nft back and forth. If the transfer does not work at all, the `then` logic will pass again so we are not testing the `transfer` this way. 
We are simply testing the `getAccountBalance`, `getTokenNftInfo`, `getAccountNftInfos` and `getTokenNftInfos` queries for mint. Now that that's the case lets remove the crypto transfers form this particular test, move the test to `management` specs (since we test the mint) and add new tests for the transfers. We have a lot of edge cases to cover there

**Review by joan41868 — APPROVED** [review]
Lookin' good

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java [hunk 19]
```diff
@@ -587,6 +593,82 @@ public HapiApiSpec updateHappyPath() {
...
+								.tokenType(TokenType.NON_FUNGIBLE_UNIQUE)
+								.treasury(TOKEN_TREASURY)
+								.initialSupply(0)
+								.adminKey("adminKeyA")
+								.supplyKey("supplyKeyA"),
+						tokenCreate("secondary")
+								.treasury("newTokenTreasury")
+								.initialSupply(0)
+								.adminKey("adminKeyB")
+								.supplyKey("supplyKeyB"),
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java:645]: Why do you create another token?

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java [hunk 20]
```diff
@@ -587,6 +593,82 @@ public HapiApiSpec updateHappyPath() {
...
+				.then(
+						getAccountBalance(TOKEN_TREASURY)
+								.hasTokenBalance("primary", 0),
+						getAccountBalance("newTokenTreasury")
+								.hasTokenBalance("primary", 1),
+						getTokenInfo("primary")
+								.hasTreasury("newTokenTreasury")
+								.logged(),
+						getTokenNftInfo("primary", 1)
+								.hasAccountID(TOKEN_TREASURY) // still the same treasury since no token transfer is initiated
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java:662]: `getTokenNftInfo` should return the new `treasury`. This is should not happen (returning TOKEN_TREASURY instead of `newTokenTreasury`)

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java [hunk 21]
```diff
@@ -587,6 +593,82 @@ public HapiApiSpec updateHappyPath() {
...
+								.logged(),
+						getTokenNftInfos("primary", 0, 1)
+								.hasNfts(
+										newTokenNftInfo("primary", 1, TOKEN_TREASURY, ByteString.copyFromUtf8("memo1"))
+								) // still the same treasury since no token transfer is initiated
+								.logged(),
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java:668]: This is should not happen (returning TOKEN_TREASURY instead of newTokenTreasury)

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
Looks good. So far I don;t see any concerns. We should resolve the complex query for the account nfts and make sure that we cover all edge cases in E2E tests and unit tests. 

We should add a lot of e2e tests for the different cases in update/mint/burn/wipe/transfer and each of them should verify that the queries are returning correct results

File: hedera-node/src/main/java/com/hedera/services/ledger/HederaLedger.java (hunk 6 above)
Comment (daniel-danev-limechain) [reply hedera-node/src/main/java/com/hedera/services/ledger/HederaLedger.java:245]: L236 is for nfts, L244 is for fungibles. This is the correct code since we need to calculate first nft and then fungibles

File: hedera-node/src/test/java/com/hedera/services/state/initialization/ViewBuilderTest.java (hunk 15 above)
Comment (daniel-danev-limechain) [reply hedera-node/src/test/java/com/hedera/services/state/initialization/ViewBuilderTest.java:94]: The unit test setup is not completed for this actualTUtao

File: hedera-node/src/test/java/com/hedera/services/store/TypedTokenStoreTest.java (hunk 16 above)
Comment (daniel-danev-limechain) [reply hedera-node/src/test/java/com/hedera/services/store/TypedTokenStoreTest.java:300]: This is never called since during minting op we do association for uniqueTreasuryOwnership FCOTMR and not for uniqueTokenOwnerships

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java (hunk 20 above)
Comment (daniel-danev-limechain) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java:663]: During tokenUpdate, nft owner is not updated to 0.0.0 and consequal execution of getTokenNftInfo and/or getTokenNftInfos retuns nft/s previous tresury as implemented by design

File: hedera-node/src/main/java/com/hedera/services/state/initialization/ViewBuilder.java [hunk 22]
```diff
@@ -23,18 +23,26 @@
...
 public class ViewBuilder {
 	public static void rebuildUniqueTokenViews(
+			TokenStore tokenStore,
 			FCMap<MerkleUniqueTokenId, MerkleUniqueToken> uniqueTokens,
 			FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueTokenAssociations,
-			FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueOwnershipAssociations
+			FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueOwnershipAssociations,
+			FCOneToManyRelation<EntityId, MerkleUniqueTokenId> treasuryOwnedUniqueTokens
 	) {
+		// call first ctx.rebuildStoreViewsIfPresent() before rebuilding the unique token views
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/state/initialization/ViewBuilder.java:38]: I was expecting this to be documented as a property java doc

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java [hunk 23]
```diff
@@ -587,6 +594,75 @@ public HapiApiSpec updateHappyPath() {
...
+						getTokenInfo("primary")
+								.hasTreasury("newTokenTreasury")
+								.logged(),
+						getTokenNftInfo("primary", 1)
+								.hasAccountID("newTokenTreasury") // still the same treasury since no token transfer is initiated
+								.logged(),
+						getTokenNftInfos("primary", 0, 1)
+								.hasNfts(
+										newTokenNftInfo("primary", 1, "newTokenTreasury", ByteString.copyFromUtf8("memo1"))
+								) // still the same treasury since no token transfer is initiated
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java:661]: remove unnecessary comments

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java [hunk 24]
```diff
@@ -482,6 +486,8 @@ public HapiApiSpec uniqueTokenTxnAccountBalance() {
 								.hasTokenBalance(A_TOKEN, 0),
 						getAccountBalance(FIRST_USER)
 								.hasTokenBalance(A_TOKEN, 1),
+						getTokenInfo(A_TOKEN)
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java:489]: Why do you have check for the treasury if you only created the token and haven't updated the treasury? Isn't it unecessary

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java [hunk 25]
```diff
@@ -496,6 +502,51 @@ public HapiApiSpec uniqueTokenTxnAccountBalance() {
...
+								.treasury(TOKEN_TREASURY),
+						mintToken(A_TOKEN, List.of(ByteString.copyFromUtf8("memo"))),
+						tokenAssociate("newTreasury", A_TOKEN)
+				).when(
+						cryptoTransfer(
+								movingUnique(1, A_TOKEN).between(TOKEN_TREASURY, "newTreasury")
+						).signedBy("signingKeyTreasury", DEFAULT_PAYER).via("cryptoTransferTxn"),
+						cryptoTransfer(
+								movingUnique(1, A_TOKEN).between("newTreasury", TOKEN_TREASURY)
+						).signedBy("signingKeyTreasury", DEFAULT_PAYER).via("cryptoTransferTxn")
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java:526]: I am not sure why we are changing the nft back and forth. If the transfer does not work at all, the then logic will pass again so we are not testing the transfer this way.
We are simply testing the getAccountBalance, getTokenNftInfo, getAccountNftInfos and getTokenNftInfos queries for mint. Now that that's the case lets remove the crypto transfers form this particular test, move the test to management specs (since we test the mint) and add new tests for the transfers. We have a lot of edge cases to cover there

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 26]
```diff
@@ -322,7 +326,7 @@ public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
 			}
 
 			final var owner = (EntityId) nftsLedger.get(nftId, OWNER);
-			if (!owner.matches(from)) {
+			if (!owner.matches(from) && !owner.matches(AccountID.getDefaultInstance())) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java]: When doing `Nft` transfers, the protobuf will not contain `0.0.0` for sender/receiver when the treasury transfers or receives. It will contain the treasury account, so it should not be the case that we need to do this check: `!owner.matches(AccountID.getDefaultInstance())`

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java [hunk 27]
```diff
@@ -496,6 +501,51 @@ public HapiApiSpec uniqueTokenTxnAccountBalance() {
 				);
 	}
 
+	public HapiApiSpec uniqueTokenTxnAccountBalancesForTreasury() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java:504]: This test is not cleanly written. Adding this as comment for the 3rd time

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

**Comment by Daniel-K-Ivanov** [comment]
We should resolve the complex query for the account nfts and make sure that we cover all edge cases in E2E tests and unit tests.

We should add a lot of e2e tests for the different cases in update/mint/burn/wipe/transfer and each of them should verify that the queries are returning correct results

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 28]
```diff
@@ -536,24 +561,76 @@ public boolean scheduleExists(ScheduleID id) {
...
+			// uniqueTokenTreasury calculations
+			if (uniqueOwnershipCountEnd >= end) {
+				if (uniqueOwnerShipCountStart < end) {
+					if (uniqueOwnerShipCountStart > start) {
+						rangeStart = 0;
+						rangeEnd = end - uniqueOwnerShipCountStart;
+					} else {
+						rangeStart = start - uniqueOwnerShipCountStart;
+						rangeEnd = end - uniqueOwnerShipCountStart;
+					}
```
Comment (joan41868) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:603]: nit: Can we just move `rangeEnd = end - uniqueOwnerShipCountStart` at the bottom of the upper `if` ?

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java (hunk 27 above)
Comment (daniel-danev-limechain) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenTransactSpecs.java:680]: Test has been refactored

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetAccountNftInfos.java [hunk 29]
```diff
@@ -37,6 +37,7 @@
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Comparator;
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetAccountNftInfos.java:40]: nit: unnecessary change

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoTransfer.java [hunk 30]
```diff
@@ -393,12 +396,19 @@ protected HapiCryptoTransfer self() {
...
 				.map(p -> p.specializedForNft(spec))
-				.collect(groupingBy(
+				.collect(Collectors.toMap(
 						TokenTransferList::getToken,
-						flatMapping(xfers -> xfers.getNftTransfersList().stream(), toList())));
+						TokenTransferList::getNftTransfersList,
+						(left, right) -> Stream.of(left, right).flatMap(Collection::stream).collect(toList()),
+						LinkedHashMap::new));
+		if (aggregated.size() != 0 && uniqueCount != aggregated.size()) {
+			throw new RuntimeException("Duplicated token key!");
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoTransfer.java:410]: I think that it would be better if we add a more descriptive message. If you see "Duplicated token key!" in the e2e you wouldn't know what exactly went wrong. Maybe a better message could be "Duplicated Token Id set in unique token movement"

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java [hunk 31]
```diff
@@ -587,6 +593,75 @@ public HapiApiSpec updateHappyPath() {
...
+						getAccountBalance("newTokenTreasury")
+								.hasTokenBalance("primary", 500),
+						getTokenInfo("primary")
+								.hasTreasury("newTokenTreasury")
+								.hasTotalSupply(500),
+						getTxnRecord("tokenUpdateTxn").logged()
+				);
+	}
+
+	public HapiApiSpec updateNftTreasuryHappyPath() {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java:626]: I think that a good test is verifying that if the new treasury has NFT balance, the `TokenUpdate` is not performed and the response code is `TRANSACTION_REQUIRES_ZERO_TOKEN_BALANCES`

**Review by Daniel-K-Ivanov — APPROVED** [review]
Good work! Added small comments, nits. Otherwise LGTM!

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
.

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 32]
```diff
@@ -536,24 +562,73 @@ public boolean scheduleExists(ScheduleID id) {
...
-							.setCreationTime(nft.getCreationTime().toGrpc())
-							.setNftID(NftID.newBuilder()
-									.setTokenID(nftId.tokenId().toGrpcTokenId())
-									.setSerialNumber(nftId.serialNumber())
-									.build())
-							.setMetadata(ByteString.copyFrom(nft.getMetadata()))
-							.build()
-			);
-		});
+		var uniqueTokenAccountMap = uniqueTokenAccountOwnerships.get();
```
[the remark this answers is not in the document]
Comment (daniel-danev-limechain) [reply hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:565]: infoForAccountNfts() method is now with nearly 100% coverage. More unit tests can be added

**Review by Daniel-K-Ivanov — APPROVED** [review]
LGTM
